docs-devsite/firestore_.firestoredataconverter.md

```diff
@@ -98,43 +98,162 @@ toFirestore(modelObject: PartialWithFieldValue<AppModelType>, options: SetOption
-  },
-  fromFirestore(
-    snapshot: QueryDocumentSnapshot,
-    options: SnapshotOptions
-  ): Post {
-    const data = snapshot.data(options) as PostDbModel;
-    return new Post(data.title, data.author);
-  }
-};
+// The `PostConverter` implements `FirestoreDataConverter<AppModelType, DbModelType>`
+// and specifies how the Firestore SDK can convert `Post` objects to `PostDbModel`
+// objects and vice versa.
+class PostConverter implements FirestoreDataConverter<Post, PostDbModel> {
+    toFirestore(post: WithFieldValue<Post>) {
+        return {
+            ttl: post.title,
+            aut: this._autFromAuthor(post.author),
+            lut: this._lutFromLastUpdatedMillis(post.lastUpdatedMillis)
+        };
+    }
+
+    fromFirestore(snapshot: QueryDocumentSnapshot, options: SnapshotOptions) {
+        const data = snapshot.data(options) as PostDbModel;
+        const author = `${data.aut.firstName} ${data.aut.lastName}`;
+        return new Post(data.ttl, author, data.lut.toMillis());
+    }
+
+    _autFromAuthor(
+        author: string | FieldValue
+    ): { firstName: string; lastName: string } | FieldValue {
+        if (typeof author !== 'string') {
+            // `author` is a FieldValue, so just return it.
+            return author;
+        }
+        const [firstName, lastName] = author.split(' ');
+        return {firstName, lastName};
+    }
+
+    _lutFromLastUpdatedMillis(
+        lastUpdatedMillis: number | FieldValue
+    ): Timestamp | FieldValue {
+        if (typeof lastUpdatedMillis !== 'number') {
+            // `lastUpdatedMillis` must be a FieldValue, so just return it.
+            return lastUpdatedMillis;
+        }
+        return Timestamp.fromMillis(lastUpdatedMillis);
+    }
+}
 
-const postSnap = await firebase.firestore()
-  .collection('posts')
-  .withConverter(postConverter)
-  .doc().get();
-const post = postSnap.data();
-if (post !== undefined) {
-  post.title; // string
-  post.toString(); // Should be defined
-  post.someNonExistentProperty; // TS error
+async function advancedDemo(db: Firestore): Promise<void> {
+    // Create a `DocumentReference` with a `FirestoreDataConverter`.
+    const documentRef = doc(db, 'posts/post123').withConverter(new PostConverter());
+
+    // The `data` argument specified to `setDoc()` is type checked by the
+    // TypeScript compiler to be compatible with `Post`. Since the `data`
+    // argument is typed as `WithFieldValue<Post>` rather than just `Post`,
+    // this allows properties of the `data` argument to also be special
+    // Firestore values that perform server-side mutations, such as
+    // `arrayRemove()`, `deleteField()`, and `serverTimestamp()`.
+    await setDoc(documentRef, {
+        title: 'My Life',
+        author: 'Foo Bar',
+        lastUpdatedMillis: serverTimestamp()
+    });
+
+    // The TypeScript compiler will fail to compile if the `data` argument to
+    // `setDoc()` is _not_ be compatible with `WithFieldValue<Post>`. This
```

Comment:
Remove "be": "...is _not_ compatible..."

Reply:
Thanks